File: native-migration/slices/01-order-tracking/ios-native/TrackingRepository.swift
import Foundation
import Combine

/// Phase 2 Native Slice: Order Tracking Repository
/// Orchestrates API calls, handles mapping to domain models, and manages polling.
class TrackingRepository {
    private let trackingApi: TrackingApiProtocol
    
    init(trackingApi: TrackingApiProtocol = TrackingApi()) {
        self.trackingApi = trackingApi
    }
    
    /// Fetches tracking data continuously every `intervalMs`.
    func streamOrderTracking(orderId: String, intervalSeconds: TimeInterval = 10.0) -> AnyPublisher<OrderTrackingResponseDTO, Error> {
        return Timer.publish(every: intervalSeconds, on: .main, in: .common)
            .autoconnect()
            .prepend(Date()) // Trigger immediately on subscribe
            .flatMap { [weak self] _ -> AnyPublisher<OrderTrackingResponseDTO, Error> in
                guard let self = self else {
                    return Fail(error: URLError(.cancelled)).eraseToAnyPublisher()
                }
                return self.trackingApi.getTracking(orderId: orderId)
            }
            .eraseToAnyPublisher()
    }
    
    /// Fetches guest tracking data once.
    func fetchGuestTracking(orderId: String, phone: String) -> AnyPublisher<OrderTrackingResponseDTO, Error> {
        return trackingApi.getGuestTracking(orderId: orderId, phone: phone)
    }
}

File: native-migration/slices/01-order-tracking/android-native/TrackingRepository.kt
package com.bhojanos.app.tracking.data

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.delay

/**
 * Phase 2 Native Slice: Order Tracking Repository
 * Orchestrates API calls, handles mapping to domain models, and manages polling.
 */
class TrackingRepository(
    private val trackingApi: TrackingApi
) {
    /**
     * Fetches tracking data continuously every [intervalMs].
     */
    fun streamOrderTracking(orderId: String, intervalMs: Long = 10000L): Flow<Result<OrderTrackingResponseDTO>> = flow {
        while (true) {
            try {
                val response = trackingApi.getTracking(orderId)
                if (response.isSuccessful && response.body() != null) {
                    emit(Result.success(response.body()!!))
                } else {
                    emit(Result.failure(Exception("API Error: ${response.code()}")))
                }
            } catch (e: Exception) {
                emit(Result.failure(e))
            }
            delay(intervalMs)
        }
    }

    /**
     * Fetches guest tracking data.
     */
    suspend fun fetchGuestTracking(orderId: String, phone: String): Result<OrderTrackingResponseDTO> {
        return try {
            val response = trackingApi.getGuestTracking(orderId, phone)
            if (response.isSuccessful && response.body() != null) {
                Result.success(response.body()!!)
            } else {
                Result.failure(Exception("API Error: ${response.code()}"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}
